domain/relation/service/relation.go

```diff
@@ -133,6 +148,85 @@ type State interface {
 	WatcherApplicationSettingsNamespace() string
 }
 
+// LeadershipService provides the API for working with the statuses of applications
+// and units, including the API handlers that require leadership checks.
+type LeadershipService struct {
+	*Service
+	leaderEnsurer leadership.Ensurer
+}
+
+func NewLeadershipService(
+	st State,
+	leaderEnsurer leadership.Ensurer,
+	logger logger.Logger,
+) *LeadershipService {
+	return &LeadershipService{
+		Service:       NewService(st, logger),
+		leaderEnsurer: leaderEnsurer,
+	}
+}
+
+// GetLocalRelationApplicationSettings returns the application settings
+// for the given application and relation identifier combination.
+// ApplicationSettings may only be read by the application leader.
+//
+// The following error types can be expected to be returned:
+//   - [corelease.ErrNotHeld] if the unit is not the leader.
+//   - [relationerrors.ApplicationNotFoundForRelation] is returned if the
+//     application is not part of the relation.
+//   - [relationerrors.RelationNotFound] is returned if the relation UUID
+//     is not found.
+func (s *LeadershipService) GetLocalRelationApplicationSettings(
+	ctx context.Context,
+	unitName unit.Name,
+	relationUUID corerelation.UUID,
+	applicationID application.ID,
+) (map[string]string, error) {
+	if err := unitName.Validate(); err != nil {
+		return nil, errors.Capture(err)
+	}
+	if err := relationUUID.Validate(); err != nil {
+		return nil, errors.Errorf(
+			"%w:%w", relationerrors.RelationUUIDNotValid, err)
+	}
+	if err := applicationID.Validate(); err != nil {
+		return nil, errors.Errorf(
+			"%w:%w", relationerrors.ApplicationIDNotValid, err)
+	}
+	settings := make(map[string]string)
+	err := s.leaderEnsurer.WithLeader(ctx, unitName.Application(), unitName.String(),
```

Comment:
todo: If the unit is not the leader, `corelease.ErrNotHeld` is returned. Update the caller to return `apiservererrors.ErrPerm` in that case. The same holds true for every uniter facade a domain service method requiring leadership.